env0/data_environment_test.go

```diff
@@ -0,0 +1,107 @@
+package env0
+
+import (
+	"regexp"
+	"testing"
+
+	"github.com/env0/terraform-provider-env0/client"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+)
+
+func TestEnvironmentDataSource(t *testing.T) {
+	environment := client.Environment{
+		Id:   "id-0",
+		Name: "my-environment-1",
+	}
+
+	otherEnvironment := client.Environment{
+		Id:   "other-id",
+		Name: "other-name",
+	}
+
+	environmentFieldsByName := map[string]interface{}{"name": environment.Name}
+	environmentFieldsById := map[string]interface{}{"id": environment.Id}
+
+	resourceType := "env0_environment"
+	resourceName := "test_environment"
+	accessor := dataSourceAccessor(resourceType, resourceName)
+
+	getValidTestCase := func(input map[string]interface{}) resource.TestCase {
+		return resource.TestCase{
+			Steps: []resource.TestStep{
+				{
+					Config: dataSourceConfigCreate(resourceType, resourceName, input),
+					Check: resource.ComposeAggregateTestCheckFunc(
+						resource.TestCheckResourceAttr(accessor, "id", environment.Id),
+						resource.TestCheckResourceAttr(accessor, "name", environment.Name),
+					),
+				},
+			},
+		}
+	}
+
+	getErrorTestCase := func(input map[string]interface{}, expectedError string) resource.TestCase {
+		return resource.TestCase{
+			Steps: []resource.TestStep{
+				{
+					Config:      dataSourceConfigCreate(resourceType, resourceName, input),
+					ExpectError: regexp.MustCompile(expectedError),
+				},
+			},
+		}
+	}
+
+	mockGetEnvironmentCall := func(returnValue client.Environment) func(mockFunc *client.MockApiClientInterface) {
+		return func(mock *client.MockApiClientInterface) {
+			mock.EXPECT().Environment(environment.Id).AnyTimes().Return(returnValue, nil)
+		}
+	}
+
+	mockListEnvironmentsCall := func(returnValue []client.Environment) func(mockFunc *client.MockApiClientInterface) {
+		return func(mock *client.MockApiClientInterface) {
+			mock.EXPECT().Environments().AnyTimes().Return(returnValue, nil)
+		}
+	}
+
+	t.Run("By ID", func(t *testing.T) {
+		runUnitTest(t,
+			getValidTestCase(environmentFieldsById),
+			mockGetEnvironmentCall(environment),
+		)
+	})
+
+	t.Run("By Name", func(t *testing.T) {
+		runUnitTest(t,
+			getValidTestCase(environmentFieldsByName),
+			mockListEnvironmentsCall([]client.Environment{environment, otherEnvironment}),
+		)
+	})
+
+	t.Run("Throw error when no name or id is supplied", func(t *testing.T) {
+		runUnitTest(t,
+			getErrorTestCase(map[string]interface{}{}, "one of `id,name` must be specified"),
+			func(mock *client.MockApiClientInterface) {},
+		)
+	})
```

Comment:
Manual assertion should be deleted but it'd be great if we can still assert that schema validation throws

Reply:
it still works 👍🏻